File: doc/api-examples/python/enable-unit.py
#!/usr/bin/python3
#
# Copyright Contributors to the Eclipse BlueChi project
#
# SPDX-License-Identifier: MIT-0

import sys
from collections import namedtuple

import dasbus.connection

bus = dasbus.connection.SystemMessageBus()

EnabledServiceInfo = namedtuple(
    "EnabledServicesInfo", ["op_type", "symlink_file", "symlink_dest"]
)
EnableResponse = namedtuple(
    "EnableResponse", ["carries_install_info", "enabled_services_info"]
)

if len(sys.argv) < 2:
    print(f"Usage: {sys.argv[0]} node_name unit_name...")
    sys.exit(1)

node_name = sys.argv[1]

controller = bus.get_proxy("org.eclipse.bluechi", "/org/eclipse/bluechi")
node_path = controller.GetNode(node_name)
node = bus.get_proxy("org.eclipse.bluechi", node_path)

response = node.EnableUnitFiles(sys.argv[2:], False, False)
enable_response = EnableResponse(*response)
if enable_response.carries_install_info:
    print("The unit files included enablement information")
else:
    print("The unit files did not include any enablement information")

for e in enable_response.enabled_services_info:
    enabled_service_info = EnabledServiceInfo(*e)
    if enabled_service_info.op_type == "symlink":
        print(
            f"Created symlink {enabled_service_info.symlink_file} -> {enabled_service_info.symlink_dest}"
        )
    elif enabled_service_info.op_type == "unlink":
        print(f'Removed "{enabled_service_info.symlink_file}".')

File: doc/api-examples/rust/enable-unit.rs
// Copyright Contributors to the Eclipse BlueChi project
//
// SPDX-License-Identifier: MIT-0

use clap::Parser;
use dbus::blocking::Connection;
use dbus::Path;
use std::time::Duration;

#[derive(Parser)]
struct Cli {
    /// The node name to list the units for
    #[clap(short, long)]
    node_name: String,

    /// The names of the units to enable. Names are separated by ','.
    #[clap(short, long, value_delimiter = ',')]
    unit_names: Vec<String>,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Cli::parse();

    let conn = Connection::new_system()?;

    let bluechi = conn.with_proxy(
        "org.eclipse.bluechi",
        "/org/eclipse/bluechi",
        Duration::from_millis(5000),
    );

    let (node,): (Path,) =
        bluechi.method_call("org.eclipse.bluechi.Controller", "GetNode", (args.node_name,))?;

    let node_proxy = conn.with_proxy("org.eclipse.bluechi", node, Duration::from_millis(5000));

    let (carries_install_info, changes): (bool, Vec<(String, String, String)>) = node_proxy
        .method_call(
            "org.eclipse.bluechi.Node",
            "EnableUnitFiles",
            (args.unit_names, false, false),
        )?;

    if carries_install_info {
        println!("The unit files included enablement information");
    } else {
        println!("The unit files did not include any enablement information");
    }

    for (op_type, file_name, file_dest) in changes {
        if op_type == "symlink" {
            println!("Created symlink {} -> {}", file_name, file_dest);
        } else if op_type == "unlink" {
            println!("Removed '{}'", file_name);
        }
    }

    Ok(())
}
